src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs

```diff
@@ -910,6 +920,49 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
             return null;
         }
 
+        public virtual void VisitPattern(BoundExpression expression, BoundPattern pattern)
+        {
+            Split();
+            bool? knownMatch = CheckRefutations(expression, pattern);
+            switch (knownMatch)
+            {
+                case true:
+                    SetState(StateWhenTrue);
+                    SetConditionalState(this.State, UnreachableState());
+                    break;
+                case false:
+                    SetState(StateWhenFalse);
+                    SetConditionalState(UnreachableState(), this.State);
+                    break;
+                case null:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Check if the given expression is known to *always* match, or *always* fail against the given pattern.
+        /// Return true for known match, false for known fail, and null otherwise.
+        /// </summary>
+        private bool? CheckRefutations(BoundExpression expression, BoundPattern pattern)
+        {
+            switch (pattern.Kind)
+            {
+                case BoundKind.DeclarationPattern:
+                    {
+                        var declPattern = (BoundDeclarationPattern)pattern;
+                        if (declPattern.IsVar || // var pattern always matches
+                            declPattern.DeclaredType?.Type?.IsValueType == true && declPattern.DeclaredType.Type == (object)expression.Type) // exact match
```

Comment:
Should type comparison here pay attention to dynamic, custom attributes, tuple names, array sizes?